test/unit/controllers/declaration/ItemTypeControllerSpec.scala

```diff
@@ -0,0 +1,245 @@
+          ("taricAdditionalCode[]", ""),
+          ("nationalAdditionalCode[]", ""),
+          ("descriptionOfGoods", "description"),
+          ("cusCode", ""),
+          ("unDangerousGoodsCode", ""),
+          ("statisticalValue", "value"),
+          ("WrongAction", "")
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(wrongAction: _*))
+
+        status(result) mustBe BAD_REQUEST
+        verify(mockItemTypePage, times(0)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "form is incorrect during adding" in {
+
+        withNewCaching(aDeclaration(withItem(anItem(withItemId(itemId)))))
+
+        val incorrectForm = Seq(
+          ("combinedNomenclatureCode", "!@#$$%"),
+          ("taricAdditionalCode[]", "!@#$$%"),
+          ("nationalAdditionalCode[]", "!@#$$%"),
+          ("descriptionOfGoods", "!@#$$%"),
+          ("cusCode", "!@#$$%"),
+          ("unDangerousGoodsCode", "!@#$$%"),
+          ("statisticalValue", "!@#$$%"),
+          addActionUrlEncoded
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(incorrectForm: _*))
+
+        status(result) mustBe BAD_REQUEST
+        verify(mockItemTypePage, times(1)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "form is incorrect during saving" in {
+
+        withNewCaching(aDeclaration(withItem(anItem(withItemId(itemId)))))
+
+        val incorrectForm = Seq(
+          ("combinedNomenclatureCode", ""),
+          ("taricAdditionalCode[]", "!@#$$%"),
+          ("nationalAdditionalCode[]", ""),
+          ("descriptionOfGoods", ""),
+          ("cusCode", ""),
+          ("unDangerousGoodsCode", ""),
+          ("statisticalValue", ""),
+          saveAndContinueActionUrlEncoded
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(incorrectForm: _*))
+
+        status(result) mustBe BAD_REQUEST
+        verify(mockItemTypePage, times(1)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+    }
+
+    "return 303 (SEE_OTHER)" when {
+
+      "item doesn't exists in the cache" in {
+
+        withNewCaching(aDeclaration())
+
+        val result = controller.displayPage(itemId)(getRequest())
+
+        status(result) mustBe SEE_OTHER
+        verify(mockItemTypePage, times(0)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "correct item is added and declaration model exist in the cache" in {
+
+        withNewCaching(aDeclaration(withItem(anItem(withItemId(itemId)))))
+
+        val correctForm = Seq(
+          ("combinedNomenclatureCode", "code"),
+          ("taricAdditionalCode[]", ""),
+          ("nationalAdditionalCode[]", ""),
+          ("descriptionOfGoods", "description"),
+          ("cusCode", ""),
+          ("unDangerousGoodsCode", ""),
+          ("statisticalValue", "1234"),
+          addActionUrlEncoded
```

Comment:
Why is there supposed to be 303 code if the action is for adding? Should it not the same page be refreshed?

Reply:
```
private def handleAddition(
    itemId: String,
    itemTypeInput: ItemType,
    itemTypeCache: ItemType,
    hasFiscalReferences: Boolean
  )(implicit request: JourneyRequest[AnyContent]): Future[Result] = {
    val itemTypeUpdated = updateCachedItemTypeAddition(itemId, itemTypeInput, itemTypeCache)
    ItemTypeValidator.validateOnAddition(itemTypeUpdated) match {
      case Valid =>
        updateExportsCache(itemId, itemTypeUpdated).map {
          case Some(model) =>
            refreshPage(itemId, itemTypeInput, model)
          case None =>
            Redirect(routes.ItemsSummaryController.displayPage())
        }
      case Invalid(errors) =>
        val formWithErrors =
          errors.foldLeft(ItemType.form().fill(itemTypeInput))((form, error) => form.withError(adjustErrorKey(error)))
        Future.successful(
          BadRequest(
            itemTypePage(
              itemId,
              adjustDataKeys(formWithErrors),
              hasFiscalReferences,
              itemTypeCache.taricAdditionalCodes,
              itemTypeCache.nationalAdditionalCodes
            )
          )
        )
    }
  }
```

Reply:
status it's OK (200) I don't understand your question

Reply:
sorted, thanks :)